Source/Core/DolphinQt/Config/VerifyWidget.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt/Config/VerifyWidget.h"
+
+#include <memory>
+#include <vector>
+
+#include <QByteArray>
+#include <QHBoxLayout>
+#include <QHeaderView>
+#include <QLabel>
+#include <QProgressDialog>
+#include <QPushButton>
+#include <QVBoxLayout>
+
+#include "Common/CommonTypes.h"
+#include "DiscIO/Volume.h"
+#include "DiscIO/VolumeVerifier.h"
+
+VerifyWidget::VerifyWidget(std::shared_ptr<DiscIO::Volume> volume) : m_volume(std::move(volume))
+{
+  QVBoxLayout* layout = new QVBoxLayout(this);
+
+  m_problems = new QTableWidget(0, 2, this);
+  m_problems->setHorizontalHeaderLabels({tr("Problem"), tr("Severity")});
+  m_problems->horizontalHeader()->setSectionResizeMode(0, QHeaderView::Stretch);
+  m_problems->horizontalHeader()->setSectionResizeMode(1, QHeaderView::ResizeToContents);
+  m_problems->verticalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);
+  m_problems->verticalHeader()->hide();
+
+  m_summary_text = new QTextEdit(this);
+  m_summary_text->setReadOnly(true);
+
+  QFormLayout* hash_layout = new QFormLayout(this);
+  CreateHashLine(hash_layout, tr("CRC32:"), &m_crc32_checkbox, &m_crc32_line_edit);
+  CreateHashLine(hash_layout, tr("MD5:"), &m_md5_checkbox, &m_md5_line_edit);
+  CreateHashLine(hash_layout, tr("SHA-1:"), &m_sha1_checkbox, &m_sha1_line_edit);
+
+  QPushButton* verify_button = new QPushButton(tr("Verify Integrity"));
+  connect(verify_button, &QPushButton::clicked, this, &VerifyWidget::Verify);
+
+  layout->addWidget(m_problems);
+  layout->addWidget(m_summary_text);
+  layout->addLayout(hash_layout);
+  layout->addWidget(verify_button);
+
+  layout->setStretchFactor(m_problems, 5);
+  layout->setStretchFactor(m_summary_text, 2);
+
+  setLayout(layout);
+}
+
+void VerifyWidget::CreateHashLine(QFormLayout* layout, QString text, QCheckBox** checkbox,
```

Comment:
These pointers-to-pointers somehow seem a bit icky (especially with the constant dereferencing going on).

Not sure how to salvage this though... Maybe just return the pointers as a pair or something?

*You can probably feel free to ignore this more of a personal gripe I guess*

Reply:
I can use `std::pair` and `std::tie`.